File: pe/visualstudio/PE.FSharp/Problem0017.fs
namespace PE.FSharp.Problem16

module Problem0017 =
    
    let onesDigitToWord onesDigit =
        match onesDigit with
        | 1 -> "one"
        | 2 -> "two"
        | 3 -> "three"
        | 4 -> "four"
        | 5 -> "five"
        | 6 -> "six"
        | 7 -> "seven"
        | 8 -> "eight"
        | 9 -> "nine"
        | _ -> failwith (onesDigit.ToString() + " is not a one digit number.")

    let teensToWord teenInt =
        match teenInt with
        | 10 -> "ten"
        | 11 -> "eleven"
        | 12 -> "twelve"
        | 13 -> "thirteen"
        | 14 -> "fourteen"
        | 15 -> "fifteen"
        | 16 -> "sixteen"
        | 17 -> "seventeen"
        | 18 -> "eighteen"
        | 19 -> "nineteen" 
        | _  -> failwith (teenInt.ToString() + " is not a teen.")

    let tensDigitToWord tensDigit =
        match tensDigit with
        | 2 -> "twenty"
        | 3 -> "thirty"
        | 4 -> "forty"
        | 5 -> "fifty"
        | 6 -> "sixty"
        | 7 -> "seventy"
        | 8 -> "eigthy"
        | 9 -> "ninety"
        | _ -> failwith (tensDigit.ToString() + " is not a tens digit.")

    let rec wordsFromInt n = 
        let lessThan100ButMoreThan20 twoDigits =
            let hyphenateWordNumber onesDigit =
                match onesDigit with
                | onesDigit when onesDigit < 1  -> ""
                | onesDigit when onesDigit < 10 -> "-" + wordsFromInt onesDigit
                | _                             -> failwith "not a valid hyphenate"
            
            String.concat "" [(tensDigitToWord (twoDigits / 10)); (hyphenateWordNumber (twoDigits % 10))]
        
        let lessThan1000ButMoreThan100 threeDigits =
            let lastTwoDigits twoDigits =
                match twoDigits with
                | twoDigits when twoDigits < 1  -> ""
                | _                             -> " and " + wordsFromInt twoDigits

            String.concat "" [(wordsFromInt (threeDigits / 100)); " hundred"; lastTwoDigits (threeDigits % 100)]

        match n with
        | n when n < 10     -> onesDigitToWord n
        | n when n < 20     -> teensToWord n
        | n when n < 100    -> lessThan100ButMoreThan20 n
        | n when n < 1000   -> lessThan1000ButMoreThan100 n
        | n when n = 1000   -> "one thousand"
        | _ -> failwith "error"


    let sumOfLengthsOfNumberWords listOfNumbers =
        listOfNumbers 
        |> List.map wordsFromInt
        |> List.map (fun elem -> elem.Replace("-", ""))
        |> List.map (fun elem -> elem.Replace(" ", ""))
        |> List.map String.length
        |> List.sum

File: pe/nonvisualstudio/haskell/OldWork/Implementation/Problem0017.hs
-- ghci 
-- :load C:\Users\Thomas\Documents\GitHub\haskell.practice\PE\Problem0017.hs 
-- :r 
-- :set +s

module Problem17 where
	import Data.Char

	onesDigitToWord 1 = "one"
	onesDigitToWord 2 = "two"
	onesDigitToWord 3 = "three"
	onesDigitToWord 4 = "four"
	onesDigitToWord 5 = "five"
	onesDigitToWord 6 = "six"
	onesDigitToWord 7 = "seven"
	onesDigitToWord 8 = "eight"
	onesDigitToWord 9 = "nine"
	onesDigitToWord _ = error "not a valid digit"
		
	teensToWord 10	= "ten"
	teensToWord 11 	= "eleven"
	teensToWord 12	= "twelve"
	teensToWord 13 	= "thirteen"
	teensToWord 14	= "fourteen"
	teensToWord 15 	= "fifteen"
	teensToWord 16	= "sixteen"
	teensToWord 17 	= "seventeen"
	teensToWord 18	= "eighteen"
	teensToWord 19 	= "nineteen"
	teensToWord _	= error "not a valid teen"
	
	tensDigitToWord 2 = "twenty"
	tensDigitToWord 3 = "thirty"
	tensDigitToWord 4 = "forty"
	tensDigitToWord 5 = "fifty"
	tensDigitToWord 6 = "sixty"
	tensDigitToWord 7 = "seventy"
	tensDigitToWord 8 = "eigthy"
	tensDigitToWord 9 = "ninety" 
	tensDigitToWord _ = error "not a valid tens"
	
	tensAndTeensToWord n 	| n < 20 	= teensToWord n
							| otherwise	= normalTens n 
								where 	normalTens n 	= (tensDigitToWord $ n `div` 10) ++ (onesDigit $ n `mod` 10)
										where 	onesDigit digit | digit < 1 = ""
																| otherwise = "-" ++ onesDigitToWord digit
	
	hundredsToWord n = 	(onesDigitToWord $ n `div` 100) ++ " hundred" ++ (tensDigits $ n `mod` 100)
								where tensDigits digits | digits < 1 	= ""
														| otherwise		= " and " ++ wordFromInt digits 
	
	
	wordFromInt n 	| n < 10 	= onesDigitToWord n
					| n < 100 	= tensAndTeensToWord n
					| n < 1000	= hundredsToWord n
					| n == 1000	= "one thousand"
	
	sumOfLengthOfListOfNumbersToWords listOfNumbers =
		sum
		$ map length
		$ map (filter (\elem -> elem /= '-' && elem /= ' '))
		$ map wordFromInt listOfNumbers
	
	answer = sumOfLengthOfListOfNumbersToWords [1..1000]
	
	--tests
	
	hundredsToWordTests = and
		[
			hundredsToWord 100 == "one hundred",
			hundredsToWord 101 == "one hundred and one",
			hundredsToWord 200 == "two hundred",
			hundredsToWord 878 == "eight hundred and seventy-eight"
		]
	
	tensDigitToWordTests = and
		[
			tensDigitToWord 2 == "twenty",
			tensDigitToWord 3 == "thirty",
			tensDigitToWord 4 == "forty",
			tensDigitToWord 5 == "fifty",
			tensDigitToWord 6 == "sixty",
			tensDigitToWord 7 == "seventy",
			tensDigitToWord 8 == "eigthy",
			tensDigitToWord 9 == "ninety" 
		]
	
	wordFromIntTests = and
		[
			wordFromInt 1 		== "one",
			wordFromInt 10 		== "ten",
			wordFromInt 12 		== "twelve",
			wordFromInt 99		== "ninety-nine",
			wordFromInt 100		== "one hundred",
			wordFromInt 70 		== "seventy",
			wordFromInt 1000 	== "one thousand",
			wordFromInt 998		== "nine hundred and ninety-eight"
		]
	
	tensAndTeensToWordTests = and
		[
			tensAndTeensToWord 11 == "eleven",
			tensAndTeensToWord 14 == "fourteen",
			tensAndTeensToWord 20 == "twenty",
			tensAndTeensToWord 30 == "thirty",
			tensAndTeensToWord 34 == "thirty-four"
		]
	
	teensToWordTests = and
		[
			teensToWord 10 == "ten",
			teensToWord 11 == "eleven",
			teensToWord 12 == "twelve",
			teensToWord 13 == "thirteen",
			teensToWord 14 == "fourteen",
			teensToWord 15 == "fifteen",
			teensToWord 16 == "sixteen",
			teensToWord 17 == "seventeen",
			teensToWord 18 == "eighteen",
			teensToWord 19 == "nineteen"
		]
	
	sumOfLengthOfListOfNumbersToWordsTests = and
		[
			sumOfLengthOfListOfNumbersToWords [1..3] 	== 11,
			sumOfLengthOfListOfNumbersToWords [1..5] 	== 19,
			sumOfLengthOfListOfNumbersToWords [11..12] 	== 12,
			sumOfLengthOfListOfNumbersToWords [89..90]	== 16,
			sumOfLengthOfListOfNumbersToWords [342]		== 23,
			sumOfLengthOfListOfNumbersToWords [115]		== 20
		]
			
	onesDigitToWordTests = and
		[
			onesDigitToWord 1 == "one",
			onesDigitToWord 2 == "two",
			onesDigitToWord 3 == "three",
			onesDigitToWord 4 == "four",
			onesDigitToWord 5 == "five",
			onesDigitToWord 6 == "six",
			onesDigitToWord 7 == "seven",
			onesDigitToWord 8 == "eight",
			onesDigitToWord 9 == "nine"
		]
	
	tests = and
		[
			onesDigitToWordTests,
			sumOfLengthOfListOfNumbersToWordsTests,
			teensToWordTests,
			tensAndTeensToWordTests,
			tensDigitToWordTests,
			hundredsToWordTests,
			wordFromIntTests
		]